pkg/segment/writer/agiletree.go

```diff
@@ -188,13 +188,28 @@ func (stb *StarTreeBuilder) resetNodeData() {
 		for k := range node.children {
 			delete(node.children, k)
 		}
-		node.aggValues = nil
+		if len(node.aggValues) > 0 {
+			for i := range node.aggValues {
+				node.aggValues[i].Reset()
+			}
+		}
 	}
 	stb.nodeCount = 0
 }
 
 func (stb *StarTreeBuilder) newNode() *Node {
 
+	// pre-alloc on the first one to the size of MaxAgileTree,
+	// and after that if the nodePool count exceeds then we can do the
+	// one by one extension
+	sizeNeeded := MaxAgileTreeNodeCount - len(stb.nodePool)
+	if sizeNeeded > 0 {
+		newArr := make([]*Node, sizeNeeded)
+		for i := 0; i < sizeNeeded; i++ {
+			newArr[i] = &Node{}
+		}
+		stb.nodePool = append(stb.nodePool, newArr...)
+	}
```

Comment:
nit: there's a util for this
```suggestion
stb.nodePool = toputils.ResizeSlice(stb.nodePool, MaxAgileTreeNodeCount)
```
For the case `sizeNeeded < 0` in your original code, I'm pretty sure the extra space in the slice will not get garbage collected (`ResizeSlice()` doesn't decrease the capacity, and I'm pretty sure the GC won't touch anything that's in the capacity but not the length, but I haven't been able to find documentation for this)

Same comment elsewhere